File: pkg/EarthSciAST.jl/src/compiler.jl
# ============================================================
# Choosing the compiler (API_SPEC §5.8, esm-libraries-spec §2.5.10)
# ============================================================
#
# `compiler` names the strategy that builds a problem's right-hand side. The
# vocabulary is closed and identical in every binding; this file holds Julia's
# side of it: the vocabulary, the per-tier PLAN each value expands to, the
# per-build REPORT that says which tier every rule landed on, and the refusal
# helper the strict `native` compiler raises through.
#
# The plan is the ONLY place a tier reads its own on/off state from. No
# environment variable selects an evaluation strategy: the vocabulary value is
# the whole answer, so an environment and an argument have nothing to disagree
# about (esm-libraries-spec §2.5.10). What survives in `ENV` is the tuning
# thresholds — a node budget, a function-size cap, a threading floor, an
# admission floor — each of which is a REFUSAL BOUNDARY under `native` rather
# than a fallback trigger, and two experimental tiers that ship off.

"""
    COMPILER_VOCABULARY

The closed set of `compiler` values (esm-libraries-spec §2.5.10), in the order
the spec lists them. A value outside it raises `compiler_unknown`; a value in
it that this binding cannot provide raises `compiler_unavailable`. Neither is
ever answered by building with a different compiler.
"""
const COMPILER_VOCABULARY = (:interpreter, :native, :xla, :mtk, :sympy)

# ---------------------------------------------------------------------------
# The plan: one Bool per tier, derived from the vocabulary value
# ---------------------------------------------------------------------------
#
# Every field is "this tier may run". `native` has them all true; `interpreter`
# has them all false. A field exists here for a tier only when turning it off
# changes which evaluator runs — a tuning threshold (a node budget, a cell
# floor) is a different thing and stays where it is, and so is MEMOIZATION.
#
# MEMOIZATION IS NOT A TIER. The load-time expansion memo
# (`_expand_expr_refs`) and the reference-preserving template image
# (`lower_expression_templates`) make the same program out of the same
# document; what they save is build wall time, not a tree walk per cell. So
# `interpreter` keeps both: an oracle that re-derives the program it is about
# to compare against would be a slower way to build the SAME answer, and on the
# documents it exists to run it is the difference between minutes and hours.
struct CompilerPlan
    name::Symbol          # the vocabulary value this plan implements
    strict::Bool          # refuse a rule this compiler cannot express
    # ---- kernel emission ----
    codegen::Bool                 # the RuntimeGeneratedFunction emitter
    dual_codegen::Bool            # its overflow (typemax-budget) second pass
    f64_overflow::Bool            # …also serving Float64 calls
    codegen_body_split::Bool
    cg_foreign_scratch::Bool
    cg_helper_dedup::Bool
    # ---- the array cascade ----
    stencil::Bool                 # the affine stencil tier
    array_contraction::Bool
    contraction_loop::Bool
    subtree_tbl::Bool
    state_box::Bool
    obsref::Bool
    lane_affine_key::Bool
    xeq_variant::Bool
    join_on_gate::Bool
    # ---- kernel assembly ----
    oop_merge::Bool               # kernel-class merges (both spellings)
    oop_merge_expand::Bool
    oop_batch::Bool
    direct_class_emit::Bool
    cross_eq_class_emit::Bool
    lane_intern::Bool
    xcse::Bool
    # ---- prelude cadence ----
    tiered::Bool                  # the const/time cadence tiers
    tcadence::Bool
    # ---- build-once machinery whose OFF state is a reference path ----
    intern::Bool
    setup_map_compile_once::Bool
    geom_sweep_specialize::Bool
    geom_overlap_gate::Bool
end

# Threading is deliberately NOT a plan field. A chunked kernel is
# value-identical to the serial one by construction — the chunks write disjoint
# slots — so it is not a choice of evaluator; how many cells are worth a thread
# dispatch is a tuning threshold (`ESS_THREADS_MIN_CELLS`), read on the
# right-hand-side call where the build's plan is no longer in scope.

_plan_all(name::Symbol, strict::Bool, on::Bool) =
    CompilerPlan(name, strict,
                 on, on, on, on, on, on,
                 on, on, on, on, on, on, on, on, on,
                 on, on, on, on, on, on, on,
                 on, on,
                 on, on, on, on)

"""
    _compiler_plan(compiler::Symbol) -> CompilerPlan

Expand a vocabulary value into the tier plan that implements it. Raises for the
values this binding does not provide, never substituting another one.
"""
function _compiler_plan(compiler::Symbol)
    if compiler === :native
        return _plan_all(:native, true, true)
    elseif compiler === :interpreter
        return _plan_all(:interpreter, false, false)
    elseif compiler === :xla
        # The specialty compiler that needs a HEAVY EXTERNAL DEPENDENCY: the
        # direct StableHLO emitter, which exists only with Reactant in the
        # session. `_xla_extension` raises `compiler_unavailable` when it is
        # not, which is the right failure for a value that IS in the
        # vocabulary, and the device is resolved to a client here too — a
        # mistyped `EARTHSCI_JULIA_XLA_DEVICE` is an `ArgumentError`, a GPU
        # this process does not have is `compiler_unavailable` — so neither
        # waits for a load, a flatten and a build. Everything else about the
        # lane is in compiler_xla.jl.
        # The PLAN is `native`'s: the emitter lowers the compiled tree-walk
        # intermediate representation, so the tiers that build it all run, and
        # they run strictly — a rule `native` would refuse is a rule `:xla` has
        # no program for either.
        _xla_extension()
        _xla_client(_xla_device())
        return _plan_all(:xla, true, true)
    elseif compiler === :mtk
        # VOCABULARY FIRST, AVAILABILITY SECOND. `:mtk` is a member of the
        # closed vocabulary, so a session without ModelingToolkit hears
        # `compiler_unavailable` naming what to load — never `compiler_unknown`,
        # which would say the value does not exist.
        #
        # `:mtk` is a SPECIALTY compiler: it works only for some documents, and
        # it is the one that runs EVENTS and IMPLICIT EQUATIONS (the constructs
        # CONFORMANCE_SPEC §5.39 has every other evaluator refuse). `:native`
        # remains the universally fast default with no heavy external
        # dependency; `:interpreter` remains the simple oracle that checks them.
        Base.get_extension(EarthSciAST, :EarthSciASTMTKExt) === nothing &&
            throw(SimulateError(
                "compiler=:mtk builds a ModelingToolkit `System` and needs " *
                "ModelingToolkit loaded — add `using ModelingToolkit` so the " *
                "EarthSciASTMTKExt extension activates. It is a SPECIALTY " *
                "compiler that runs only some documents (it is the one that runs " *
                "events and implicit equations); compiler=:native is the " *
                "universally fast default",
                ERROR_CODES.COMPILER_UNAVAILABLE))
        # Every `native` tier is OFF: this compiler emits nothing of this
        # package's own, so a tier flag would describe a cascade that never
        # runs. `strict` stays TRUE — `:mtk` refuses a document it cannot
        # express by name (`compiler_refused_rule`) rather than building part
        # of it, which is what `_refuse_rule` reads off the plan in force.
        return _plan_all(:mtk, true, false)
    elseif compiler === :sympy
        throw(SimulateError(
            "compiler=:sympy is a Python-binding compiler (a lambdified SymPy " *
            "scalar right-hand side) and has no Julia implementation; use " *
            "compiler=:native, or the Python binding's esm_problem(..., " *
            "compiler=\"sympy\")",
            ERROR_CODES.COMPILER_UNAVAILABLE))
    end
    throw(SimulateError(
        "compiler=:$compiler is outside the vocabulary; it is one of " *
        join((":" * String(v) for v in COMPILER_VOCABULARY), ", "),
        ERROR_CODES.COMPILER_UNKNOWN))
end

# The plan in force. Task-local so a nested build cannot leak its plan into the
# caller's, with a process-wide default for the entry points that never go
# through `_with_compiler_plan` (a direct `_compile` call in a test, say).
# `Base.ScopedValues` would say this more directly but is 1.11+, and this
# package supports 1.10. That default is NON-strict: outside a build there is
# no `compiler` keyword to have been strict about, and a refusal raised there
# would name a compiler nobody asked for.
const _COMPILER_PLAN_KEY = :earthsci_compiler_plan
const _DEFAULT_COMPILER_PLAN = _plan_all(:native, false, true)

_compiler_plan_now()::CompilerPlan =
    get(task_local_storage(), _COMPILER_PLAN_KEY, _DEFAULT_COMPILER_PLAN)

_with_compiler_plan(f, plan::CompilerPlan) =
    task_local_storage(f, _COMPILER_PLAN_KEY, plan)

# The plan a `compiler` keyword produces. The keyword has no "unset" value: a
# caller who names nothing gets the strict `:native` a caller who names it
# gets, because a default that was quietly more permissive than the same value
# spelled out is the second way of saying the same thing that §2.5.10 exists to
# remove.
_plan_for(compiler::Symbol) = _compiler_plan(compiler)

# ---------------------------------------------------------------------------
# The report: which tier each rule landed on
# ---------------------------------------------------------------------------

"""
    CompilerRuleRecord

One rule of one build: an equation, an observed, or a setup array, the tier it
landed on, and every decline it collected getting there.

* `rule` — the rule's identity, component-qualified the way the document spells
  it (a derivative target with its output axes, an observed's name, a setup
  array's name).
* `kind` — `:equation`, `:observed`, `:setup_array`, or `:rhs_program` for a
  compiler that emits ONE program for the whole assembled right-hand side
  (`:xla`) rather than lowering it rule by rule.
* `tier` — where it landed: `:affine`, `:scan`, `:array_contraction_codegen`,
  `:percell_build` (scalarized per output cell at BUILD, then compiled),
  `:codegen`, `:interpreter`, `:setup_compiled`, `:setup_percell`; and, on the
  `:rhs_program` row an `:xla` build adds, `:xla_direct_cpu` /
  `:xla_direct_gpu`, which name the emitter and the device together.
* `declines` — `tier => reason` for every tier that looked at this rule and
  passed, deepest reason last.
"""
struct CompilerRuleRecord
    rule::String
    kind::Symbol
    tier::Symbol
    declines::Vector{Pair{Symbol,Symbol}}
end

"""
    CompilerReport

What a build did, per rule. [`compiler_report`](@ref) returns one.

* `compiler` — the vocabulary value that built this problem.
* `rules` — a [`CompilerRuleRecord`](@ref) per equation / observed / setup
  array, in build order.
* `tally` — this build's cascade counters (the per-build twin of the
  process-global `_CASCADE_TALLY`), for the counts a per-rule row does not
  carry: emitted kernels, class merges, subtree-table rescues.

`tier_histogram(report)` folds `rules` into tier → count, which is what
`show(::EsmProblem)` prints.
"""
struct CompilerReport
    compiler::Symbol
    rules::Vector{CompilerRuleRecord}
    tally::Dict{Symbol,Int}
end
CompilerReport(compiler::Symbol) =
    CompilerReport(compiler, CompilerRuleRecord[], Dict{Symbol,Int}())

"""
    tier_histogram(report::CompilerReport) -> Vector{Pair{Symbol,Int}}

The report's rules folded into `tier => count`, ordered by descending count and
then by tier name, so two builds of the same document print the same line.
"""
function tier_histogram(r::CompilerReport)
    h = Dict{Symbol,Int}()
    for rec in r.rules
        h[rec.tier] = get(h, rec.tier, 0) + 1
    end
    out = collect(h)
    sort!(out; by = kv -> (-kv[2], String(kv[1])))
    return out
end

function Base.show(io::IO, r::CompilerReport)
    print(io, "CompilerReport(:", r.compiler, ", ", length(r.rules), " rules")
    for (tier, n) in tier_histogram(r)
        print(io, ", ", tier, "=", n)
    end
    print(io, ")")
end

function Base.show(io::IO, ::MIME"text/plain", r::CompilerReport)
    println(io, "CompilerReport — compiler :", r.compiler)
    hist = tier_histogram(r)
    println(io, "  tiers: ",
            isempty(hist) ? "(no rules recorded)" :
            join(("$(t)=$(n)" for (t, n) in hist), ", "))
    for rec in r.rules
        print(io, "  ", rec.kind, " ", rec.rule, " → ", rec.tier)
        isempty(rec.declines) ||
            print(io, "  [declined: ",
                  join(("$(t):$(why)" for (t, why) in rec.declines), ", "), "]")
        println(io)
    end
    return nothing
end

# ---------------------------------------------------------------------------
# The per-build record
# ---------------------------------------------------------------------------
#
# The process-global `_CASCADE_TALLY` cannot answer "what did THIS build do" —
# two builds in one session add into the same Dict, and a caller holding two
# problems cannot tell them apart. This record is per build: `build_evaluator`
# installs one for the duration of the build and the problem keeps it.
mutable struct _BuildRecord
    plan::CompilerPlan
    rules::Vector{CompilerRuleRecord}
    tally::Dict{Symbol,Int}
    # The rule the cascade is working on, so a decline raised several frames
    # deeper can name it. Empty between rules.
    current::String
    current_kind::Symbol
    declines::Vector{Pair{Symbol,Symbol}}
end
_BuildRecord(plan::CompilerPlan) =
    _BuildRecord(plan, CompilerRuleRecord[], Dict{Symbol,Int}(), "", :equation,
                 Pair{Symbol,Symbol}[])

const _BUILD_RECORD_KEY = :earthsci_build_record

_build_record()::Union{Nothing,_BuildRecord} =
    get(task_local_storage(), _BUILD_RECORD_KEY, nothing)

_with_build_record(f, rec::_BuildRecord) =
    task_local_storage(f, _BUILD_RECORD_KEY, rec)

# Name the rule the cascade is about to work on. Declines recorded while it is
# open attach to it; `_land_rule!` closes it.
function _open_rule!(rule::AbstractString, kind::Symbol)
    rec = _build_record()
    rec === nothing && return nothing
    rec.current = String(rule)
    rec.current_kind = kind
    empty!(rec.declines)
    return nothing
end

function _note_decline!(tier::Symbol, reason::Symbol)
    rec = _build_record()
    rec === nothing && return nothing
    push!(rec.declines, tier => reason)
    return nothing
end

function _land_rule!(tier::Symbol)
    rec = _build_record()
    rec === nothing && return nothing
    isempty(rec.current) && return nothing
    push!(rec.rules, CompilerRuleRecord(rec.current, rec.current_kind, tier,
                                        copy(rec.declines)))
    rec.current = ""
    empty!(rec.declines)
    return nothing
end

# A rule with no cascade stage of its own (a setup array, a materialized
# observed): one call, no open/close pair.
function _record_rule!(rule::AbstractString, kind::Symbol, tier::Symbol;
                       declines::Vector{Pair{Symbol,Symbol}} = Pair{Symbol,Symbol}[])
    rec = _build_record()
    rec === nothing && return nothing
    push!(rec.rules, CompilerRuleRecord(String(rule), kind, tier, declines))
    # Closing the rule matters as much as filing it: a label left open would be
    # read by the next refusal several stages later and name the wrong thing.
    rec.current == String(rule) && (rec.current = ""; empty!(rec.declines))
    return nothing
end

# The rule the cascade currently has open, for a refusal raised deeper than the
# stage that named it. `fallback` is what a refusal outside any single rule says
# — the assembled right-hand side belongs to every array equation at once, so
# there is no one rule to name.
_current_rule_label(fallback::AbstractString = "(unnamed rule)") =
    (rec = _build_record(); rec === nothing || isempty(rec.current) ?
     String(fallback) : rec.current)

function _finish_report(rec::_BuildRecord)
    return CompilerReport(rec.plan.name, copy(rec.rules), copy(rec.tally))
end

# ---------------------------------------------------------------------------
# The refusal
# ---------------------------------------------------------------------------

"""
    _refuse_rule(rule, reason) -> never returns

Raise `compiler_refused_rule` for the compiler in force. The message names the
compiler, the rule (component-qualified) and the deepest reason, which is the
shape §2.5.10 fixes and what the compiler-agreement tier reads back.
"""
function _refuse_rule(rule::AbstractString, reason::AbstractString)
    plan = _compiler_plan_now()
    throw(TreeWalkError(ERROR_CODES.COMPILER_REFUSED_RULE,
        "compiler=:$(plan.name) refuses '$rule': $reason"))
end

# True when the compiler in force refuses a rule it cannot express rather than
# demoting it. Every refusal site is guarded by this, so `:interpreter` — which
# is DEFINED as the slow path — never trips one.
_compiler_is_strict() = _compiler_plan_now().strict

"""
    _refuse_percell_evaluation(rule, what, cells) -> nothing or never returns

The refusal for a materialization that RESOLVES AND COMPILES the expression once
per cell. §2.5.10 puts every evaluation a compiler performs for the problem
under the refusal rule, not the right-hand side alone: the materialization of
constants and static observeds at construction, the initial-state seed and the
observeds reported at output times are each a place a binding walks the tree per
cell, and each of them makes the compiler's name describe nothing if it is
allowed to happen quietly.

A compile-once sweep that then evaluates an ALREADY COMPILED node per cell is
not one of these and never reaches here: what is refused is re-deriving the
program for every cell.
"""
function _refuse_percell_evaluation(rule::AbstractString, what::AbstractString,
                                    cells::Union{Nothing,Integer})
    _compiler_is_strict() || return nothing
    over = cells === nothing ? "" :
           ", over $cells cell" * (cells == 1 ? "" : "s")
    _refuse_rule(rule,
        "$what resolves and compiles the expression once per cell$over, " *
        "because the compile-once form declined it. That is a tree walk per " *
        "cell at construction time, which esm-libraries-spec §2.5.10 puts " *
        "under the same rule as the right-hand side. Build with " *
        "compiler=:interpreter to run it")
end

File: pkg/earthsci-ast-py/src/earthsci_ast/compiler.py
"""Which strategy builds a Problem's right-hand side (``API_SPEC.md`` §5.8,
esm-libraries-spec §2.5.10).

``esm_problem(..., compiler=...)`` names one member of a CLOSED vocabulary —
``native``, ``interpreter``, ``xla``, ``mtk``, ``sympy`` — and the Problem
reports back both the value that built it and, per rule, the tier each rule
landed on. Before this module the Python binding chose between a lambdified
SymPy form and the NumPy interpreter by inspecting the document, and then chose
among the interpreter's seven fast tiers per aggregate at evaluation time, and
neither choice was expressible or readable by a caller.

Three things live here.

**The vocabulary and its refusals.** :func:`resolve_compiler` maps the keyword
to a member or raises: ``compiler_unknown`` outside the vocabulary,
``compiler_unavailable`` for ``xla`` / ``mtk``, which this binding does not
provide. Neither is ever answered by building with a different compiler — the
point of naming one is that the caller knows which ran.

**The policy.** A :class:`CompilerPolicy` is the compiler's choice made
AMBIENT for the duration of a build: :func:`use_policy` installs one on a
:class:`~contextvars.ContextVar` that the NumPy interpreter's aggregate ladder
reads. Ambient rather than threaded through every ``EvalContext`` on purpose:
under ``native`` the refusal must fire wherever a per-cell walk is reached, and
a context that was built without the policy would refuse nothing while
reporting that ``native`` ran. An evaluation outside any :func:`use_policy`
scope — a unit test driving a synthetic context, say — sees no policy and
behaves exactly as it did before this module existed.

**The recorder.** Each tier of the ladder calls :meth:`CompilerPolicy.decline`
with its own name and the reason it could not take the node, and the tier that
answers calls :meth:`CompilerPolicy.land`. The accumulated declines travel with
the landing, so a rule that ends on the per-cell walk carries the whole chain
that put it there — which is what a ``native`` refusal names, and what
:attr:`~earthsci_ast.problem.EsmProblem.compiler_report` shows for a rule that
landed on a fast tier.
"""

from __future__ import annotations

import contextvars
from collections.abc import Iterator
from contextlib import contextmanager
from dataclasses import dataclass, field
from typing import Literal

from .error_handling import (
    COMPILER_REFUSED_RULE,
    COMPILER_UNAVAILABLE,
    COMPILER_UNKNOWN,
)
from .errors import SimulationError

#: The closed vocabulary, in the order §5.8's table lists it. Every binding has
#: the same five members; which of them it PROVIDES is per-binding.
COMPILERS: tuple[str, ...] = ("native", "interpreter", "xla", "mtk", "sympy")

#: The type a caller's ``compiler`` keyword takes.
Compiler = Literal["native", "interpreter", "xla", "mtk", "sympy"]

#: The vocabulary members Python does not provide, and what would have to exist
#: for each. Named in the refusal so the message is the ruling, not a machine's
#: configuration.
_UNAVAILABLE: dict[str, str] = {
    "xla": (
        "the Python binding has no StableHLO emitter; XLA lowering lives in the "
        "Julia binding (through Reactant) and in the Rust binding (the `xla` "
        "cargo feature), and providing it here would mean writing one"
    ),
    "mtk": (
        "ModelingToolkit is a Julia package and has no Python equivalent this "
        "binding could load; `compiler='mtk'` is available in the Julia binding "
        "only, which is also the only place events and implicit equations run"
    ),
}

#: The per-cell tiers: a Python tree walk per output cell. Under ``native``
#: reaching one of these is :data:`COMPILER_REFUSED_RULE`; under ``interpreter``
#: it is where every aggregate lands by design.
PER_CELL_TIERS: frozenset[str] = frozenset({"scalar", "scalar-gate-driven", "ragged"})


class CompilerUnknownError(SimulationError):
    """``compiler_unknown``: the value is outside the closed vocabulary."""

    code = COMPILER_UNKNOWN

    def __init__(self, value: object) -> None:
        self.value = value
        super().__init__(
            f"compiler_unknown: {value!r} is not a compiler; the closed vocabulary "
            f"(API_SPEC.md §5.8) is {', '.join(repr(c) for c in COMPILERS)}"
        )


class CompilerUnavailableError(SimulationError):
    """``compiler_unavailable``: in the vocabulary, but not provided here."""

    code = COMPILER_UNAVAILABLE

    def __init__(self, value: str, reason: str) -> None:
        self.value = value
        self.reason = reason
        super().__init__(
            f"compiler_unavailable: the Python binding cannot provide "
            f"compiler={value!r} — {reason}. A compiler this binding does not "
            f"have is refused, never substituted (esm-libraries-spec §2.5.10)."
        )


class CompilerRefusedRuleError(SimulationError):
    """``compiler_refused_rule``: this compiler cannot run this rule.

    Raised at CONSTRUCTION, naming the compiler, the rule (an equation or an
    observed, component-qualified) and the reason. Never a fallback: the rule is
    not then run on a slower path.
    """

    code = COMPILER_REFUSED_RULE

    def __init__(
        self,
        compiler: str,
        rule: str,
        reason: str,
        *,
        declines: tuple[tuple[str, str], ...] = (),
        phase: str = "construction",
    ) -> None:
        self.compiler = compiler
        self.rule = rule
        self.reason = reason
        self.declines = declines
        self.phase = phase
        chain = (
            "".join(f"\n    {tier}: {why}" for tier, why in declines)
            if declines
            else "\n    (no fast tier was attempted)"
        )
        super().__init__(
            f"compiler_refused_rule: compiler={compiler!r} cannot run {rule} "
            f"({phase}) — {reason}. The tiers that declined, fastest first:{chain}\n"
            f"  A refusal is not a fallback: {compiler!r} will not run this rule on "
            f"a slower path. Build with compiler='interpreter' to run it per cell, "
            f"or with compiler='sympy' if the document is scalar."
        )


def resolve_compiler(value: str | None) -> str:
    """The vocabulary member ``value`` names, or the refusal it earns.

    ``None`` means ``native``: the default is strict, and a binding MUST NOT
    read an unspecified compiler as permission to pick a slower one on the
    grounds that the document is small or awkward (esm-libraries-spec §2.5.10).
    """
    if value is None:
        return "native"
    if not isinstance(value, str):
        raise CompilerUnknownError(value)
    if value not in COMPILERS:
        raise CompilerUnknownError(value)
    if value in _UNAVAILABLE:
        raise CompilerUnavailableError(value, _UNAVAILABLE[value])
    return value


# --------------------------------------------------------------------------- #
# The per-rule report
# --------------------------------------------------------------------------- #


@dataclass(frozen=True)
class RuleTier:
    """One rule's landing: which tier answered, and what declined first.

    ``rule`` is the equation or observed, component-qualified, in the spelling
    the refusal would name. ``phase`` is WHEN the evaluation happened — a
    Python build evaluates rules in three places (§2.5.10 puts all three under
    the compiler), and the census found six sevenths of the per-cell landings in
    the first of them:

    * ``construction`` — the const-geometry hoist and the static-observed
      materialization, inside ``esm_problem`` itself;
    * ``rhs`` — one evaluation of the right-hand side;
    * ``observed-output`` — the observeds reported at output times.

    ``declines`` is the chain that led here, fastest tier first, as
    ``(tier, reason)`` pairs.
    """

    rule: str
    phase: str
    tier: str
    declines: tuple[tuple[str, str], ...] = ()

    @property
    def per_cell(self) -> bool:
        """Did this rule land on a Python walk per output cell?"""
        return self.tier in PER_CELL_TIERS

    def __str__(self) -> str:  # pragma: no cover - display only
        chain = " ← ".join(t for t, _ in self.declines)
        via = f" (declined: {chain})" if chain else ""
        return f"{self.rule} [{self.phase}] → {self.tier}{via}"


class CompilerReport:
    """Per rule, the tier it landed on and what declined on the way.

    One entry per aggregate LANDING, not per rule: a rule whose body holds two
    aggregates, or whose aggregate is re-evaluated in a second phase, produces
    an entry each, which is what makes the record a cost story rather than a
    summary. A rule with no aggregate at all produces no entry — the whole
    ladder this records is the ``faq`` ladder — so :meth:`rules` is the rules
    that had a tier to land on.
    """

    def __init__(self) -> None:
        self._entries: list[RuleTier] = []
        self._seen: set[RuleTier] = set()

    def add(self, entry: RuleTier) -> None:
        """Record a landing, ONCE per distinct (rule, phase, tier, chain).

        A right-hand side is evaluated thousands of times in a solve and the
        ladder's answer for a given node does not change between them, so a
        record that appended per call would grow without bound and say nothing
        the first entry did not. Deduplicating makes the report the cost story
        §5.8 describes rather than a log — which is also why the counts in
        :meth:`tiers` are DISTINCT landings, not evaluations.
        """
        if entry in self._seen:
            return
        self._seen.add(entry)
        self._entries.append(entry)

    @property
    def entries(self) -> tuple[RuleTier, ...]:
        return tuple(self._entries)

    def rules(self) -> tuple[str, ...]:
        """Every rule that landed somewhere, in first-landing order."""
        seen: dict[str, None] = {}
        for e in self._entries:
            seen.setdefault(e.rule, None)
        return tuple(seen)

    def for_rule(self, rule: str) -> tuple[RuleTier, ...]:
        return tuple(e for e in self._entries if e.rule == rule)

    def tiers(self) -> dict[str, int]:
        """Landing counts by tier — the one-line cost summary."""
        out: dict[str, int] = {}
        for e in self._entries:
            out[e.tier] = out.get(e.tier, 0) + 1
        return out

    def per_cell_rules(self) -> tuple[str, ...]:
        """The rules that walked per cell. Empty under a build that survived
        ``native``, by construction."""
        seen: dict[str, None] = {}
        for e in self._entries:
            if e.per_cell:
                seen.setdefault(e.rule, None)
        return tuple(seen)

    def __iter__(self) -> Iterator[RuleTier]:
        return iter(self._entries)

    def __len__(self) -> int:
        return len(self._entries)

    def __bool__(self) -> bool:
        return bool(self._entries)

    def __repr__(self) -> str:  # pragma: no cover - display only
        return f"CompilerReport({len(self._entries)} landings, tiers={self.tiers()})"

    def __str__(self) -> str:  # pragma: no cover - display only
        return "\n".join(str(e) for e in self._entries) or "CompilerReport: no aggregates"


# --------------------------------------------------------------------------- #
# The policy: the compiler's choice, made ambient for one build
# --------------------------------------------------------------------------- #


@dataclass
class CompilerPolicy:
    """The active compiler, its strictness, and the report it is filling.

    ``strict`` (``native``) makes a per-cell landing a
    :class:`CompilerRefusedRuleError`. ``every_tier_off`` (``interpreter``)
    sends every aggregate to the per-cell walk and turns the source codegen and
    the shape-interval hull off, so the reference evaluator is a genuinely
    second implementation rather than the fast one with a flag flipped.
    """

    compiler: str
    strict: bool = False
    every_tier_off: bool = False
    report: CompilerReport = field(default_factory=CompilerReport)
    #: The rule currently being evaluated, set by :func:`rule_scope`. ``None``
    #: means the evaluation is not attributable to one rule — the setup passes
    #: that resolve index sets and invent values — and is recorded as ``<setup>``.
    rule: str | None = None
    #: Which of §2.5.10's three evaluations is running, set by :func:`phase_scope`.
    phase: str = "construction"
    # Per-NODE decline accumulators, one frame per aggregate node currently in
    # the ladder. A STACK rather than one list because an aggregate body may hold
    # another aggregate: the inner node's ladder runs inside the outer node's
    # tier, and with one list the inner landing would consume the outer node's
    # chain. `rule_scope` clears the stack at each rule boundary, so a frame left
    # behind by an evaluation that raised cannot leak into the next rule.
    _frames: list[list[tuple[str, str]]] = field(default_factory=list)
    # Whether the whole-box tier that is about to land got its body from
    # generated source (`numpy_codegen`) rather than the compiled closure. Set
    # by `_codegen_box_fn`, read once by the landing.
    _codegen: bool = False

    # -- the ladder's hooks ------------------------------------------------- #

    def begin_node(self) -> None:
        """A fresh aggregate node enters the ladder."""
        self._frames.append([])
        self._codegen = False

    def decline(self, tier: str, reason: str) -> None:
        """``tier`` could not take the innermost node, for ``reason``."""
        if not self._frames:
            self._frames.append([])
        self._frames[-1].append((tier, reason))

    def _pop_declines(self) -> tuple[tuple[str, str], ...]:
        return tuple(self._frames.pop()) if self._frames else ()

    def note_codegen(self, used: bool) -> None:
        """Whether the box body about to be evaluated is generated source."""
        self._codegen = used

    def land(self, tier: str) -> None:
        """``tier`` answered this node. Records the landing and the chain."""
        if tier in ("broadcast", "map") and self._codegen:
            tier = "codegen"
        self.report.add(
            RuleTier(
                rule=self.rule or "<setup>",
                phase=self.phase,
                tier=tier,
                declines=self._pop_declines(),
            )
        )
        self._codegen = False

    def land_per_cell(self, tier: str) -> None:
        """A per-cell walk is about to run. Under ``native`` it never does.

        The refusal is raised HERE rather than at the ladder's exit because the
        walk is the expensive thing: a caller who asked for vectorized NumPy
        must not pay for one cell of it.
        """
        if self.strict:
            declines = self._pop_declines()
            reason = declines[-1][1] if declines else "no fast tier applies"
            raise CompilerRefusedRuleError(
                self.compiler,
                self.rule or "<setup>",
                reason,
                declines=declines,
                phase=self.phase,
            )
        self.land(tier)


_ACTIVE: contextvars.ContextVar[CompilerPolicy | None] = contextvars.ContextVar(
    "earthsci_ast_compiler_policy", default=None
)


def active_policy() -> CompilerPolicy | None:
    """The policy in force, or ``None`` outside every :func:`use_policy` scope."""
    return _ACTIVE.get()


@contextmanager
def use_policy(policy: CompilerPolicy | None) -> Iterator[CompilerPolicy | None]:
    """Install ``policy`` for the duration of the block."""
    token = _ACTIVE.set(policy)
    try:
        yield policy
    finally:
        _ACTIVE.reset(token)


@contextmanager
def rule_scope(rule: str) -> Iterator[None]:
    """Attribute every landing inside the block to ``rule``. A no-op with no
    policy installed, which is what keeps the attribution free for a caller who
    did not ask for a report."""
    policy = _ACTIVE.get()
    if policy is None:
        yield
        return
    previous = policy.rule
    policy.rule = rule
    policy._frames.clear()
    try:
        yield
    finally:
        policy.rule = previous
        policy._frames.clear()


@contextmanager
def phase_scope(phase: str) -> Iterator[None]:
    """Record every landing inside the block under ``phase``."""
    policy = _ACTIVE.get()
    if policy is None:
        yield
        return
    previous = policy.phase
    policy.phase = phase
    try:
        yield
    finally:
        policy.phase = previous


def every_tier_off() -> bool:
    """Is the reference evaluator selected? Read by the tiers that are OFF under
    ``interpreter`` but live outside the aggregate ladder — the source codegen
    and the shape-inference interval hull."""
    policy = _ACTIVE.get()
    return policy is not None and policy.every_tier_off
